HCore-Web/Providers/Impl/DownloadProcessingProxyUrlProviderImpl.cs

```diff
@@ -0,0 +1,197 @@
+            ;
+            return Task.FromResult(proxyUri.Uri);
+        }
+
+        public virtual async Task<IDownloadFileData> GetFileDataAsync(HttpRequest request, Stream inputData)
+        {
+            Uri downloadUri = new Uri(request.Query["u"]);
+            string fileName = request.Query["fileName"];
+            string mimeType = request.Query["mime"];
+            string characterSet = "utf-8";
+            string originalHash = ConvertParametersFromQueryData(request);
+
+            if (!IsHashValid(originalHash, downloadUri, fileName, mimeType))
+            {
+                throw new UnauthorizedAccessException("Invalid query data");
+            }
+
+            Stream fileData = inputData;
+            if (fileData == null)
+            {
+                var remoteRequest = new HttpRequestMessage(HttpMethod.Get, downloadUri);
+                var client = _serviceProvider.GetService<HttpClient>();
+                var response = await client.SendAsync(remoteRequest).ConfigureAwait(false);
+
+                if (response.IsSuccessStatusCode)
+                {
+                    fileData = await response.Content.ReadAsStreamAsync().ConfigureAwait(false);
+
+                    if (string.IsNullOrEmpty(mimeType))
+                    {
+                        mimeType = response.Content.Headers.ContentType.MediaType;
+                    }
+
+                    characterSet = response.Content.Headers.ContentType.CharSet;
+                }
+                else
+                {
+                    _logger.LogError(
+                        $"Failed to download file from source location {downloadUri}. " +
+                        $"file name: {JsonConvert.SerializeObject(fileName)}. " +
+                        $"HTTP result was: {response.StatusCode} {response.ReasonPhrase}, response body: {response.Content}"
+                    );
+
+                    throw new HttpRequestException("Failed to download file from source location!");
+                }
+            }
+
+            return new DownloadFileDataImpl()
+            {
+                data = fileData,
+                FileName = fileName,
+                CharacterSet = characterSet,
+                MimeType = mimeType,
+            };
+        }
+
+        public virtual Uri GetSourceUri(HttpRequest request)
+        {
+            return new Uri(request.Query["u"]);
+        }
+
+        private dynamic ConvertParametersFromQueryData(HttpRequest request)
+        {
+            if (request == null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            if (!request.Query.TryGetValue(UrlQueryKeyName, out var queryData) || queryData.Count != 1)
+            {
+                return null;
+            }
+
+            string protectedJson = queryData[0];
+            string json = _protector.Unprotect(protectedJson);
+            return DeserializeObject<dynamic>(json);
```

Comment:
Please strongly type, there is no reason for dynamic

Reply:
removed, refactored